### PR TITLE
Implement server diagnostics

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## Usage and product changes

We implement completely anonymous diagnostics reporting in TypeDB Core using Sentry. This includes sending unexpected system errors and expected user errors to the backend reporting system. We also submit occasional system status updates for number of databases/database size (daily) and periodically profile a transaction. This instrumentation should help get real-world usage and error analysis to help guide the development and efforts in TypeDB.

Error and diagnostic reporting can be completely disabled in the configuration file by setting:
```
diagnostics:
  reporting:
    enable: false
```
or booting the server with `--diagnostics.reporting.enable=false`.

## Implementation

* Add Sentry as a reporting backend, which recieves updates attached to anonymous server ID
  * Always report system errors and report user-errors in a different channel
  * Submit very basic database information daily
  * Submit one transaction's duration at most hourly
* Add configuration options for diagnostics reporting, which can be enabled or disabled (enabled by default).
* No diagnostics are sent within the first hour of system bootup, since any CI or test jobs would otherwise be reported. 

Current usage of 'diagnostic' transactions switches between no-op and real transactions using time-based windows.

### CODE REVIEW

File: common/parameters/Context.java [hunk 1]
```diff
@@ -97,6 +100,15 @@ public Transaction id(long transactionId) {
         public long id() {
             return transactionId;
         }
+
+        public Transaction diagnosticTxn(ITransaction txn) {
```
Comment (flyingsilverfin) [inline common/parameters/Context.java:104]: We attach the diagnostic txn so we can add things like query structure into from within the transaction later

File: database/RocksProperties.java [hunk 2]
```diff
@@ -215,43 +275,35 @@ public void run() {
...
                 Arrays.fill(numDeletions, 0);
-                long[] keysEstimate = new long[cfHandles.size()];
+                long[] keysEstimate = new long[reader.cfHandles.size()];
                 long sstFiles = 0;
-                for (int i = 0; i < cfHandles.size(); i++) {
-                    ColumnFamilyHandle cfHandle = cfHandles.get(i);
-                    Map<String, TableProperties> sstProperties = rocksDB.getPropertiesOfAllTables(cfHandle);
+                for (int i = 0; i < reader.cfHandles.size(); i++) {
+                    ColumnFamilyHandle cfHandle = reader.cfHandles.get(i);
+                    Map<String, TableProperties> sstProperties = reader.rocksDB.getPropertiesOfAllTables(cfHandle);
```
Comment (alexjpwalker) [inline database/RocksProperties.java:306]: Let's test that these measurements are meaningful over time.

File: database/CoreTransaction.java [hunk 3]
```diff
@@ -139,6 +142,7 @@ protected void notifyClosed() {
     abstract void delete();
 
     protected void closeResources() {
+        context.diagnosticTxn().finish();
```
Comment (alexjpwalker) [inline database/CoreTransaction.java:145]: Let's add a quick comment on the PR itself saying why NoOpTransaction is a common pattern in tracing

File: server/resources/typedb-ascii.txt [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-       |__|      |__|   |__|     |_______||_____/  |_____/
+========================================================================================================
+     ________ __      __ _____    _______  _____    _____       _______  _______  _____     _______
+    |__    __|\  \  /  /|   _  \ |   _   ||   _  \ |   _  \    |   _   ||   _   ||   _  \  |   _   |
+       |  |    \  \/  / |  | |  ||  | |__||  | |  ||  | |  |   |  | |  ||  | |  ||  | |  | |  | |__|
+       |  |     \    /  |  |/  / |  |___  |  | |  ||  |/  /    |  | |__||  | |  ||  |/  /  |  |___
+       |  |      |  |   |   __/  |   ___| |  | |  ||   _  \    |  |  __ |  | |  ||     <   |   ___|
+       |  |      |  |   |  |     |  |  __ |  | |  ||  | |  |   |  | |  ||  | |  ||  |\  \  |  |  __
+       |  |      |  |   |  |     |  |_|  ||  |/  / |  |/  /    |  |_|  ||  |_|  ||  | \  \ |  |_|  |
+       |__|      |__|   |__|     |_______||_____/  |_____/     |_______||_______||__|  \__\|_______|
```
Comment (alexjpwalker) [inline server/resources/typedb-ascii.txt:9]:
```
  _   _ _____ _____ ______ 
 | \ | |_   _/ ____|  ____|
 |  \| | | || |    | |__   
 | . ` | | || |    |  __|  
 | |\  |_| || |____| |____ 
 |_| \_|_____\_____|______|
```

File: database/RocksProperties.java (hunk 2 above)
Comment (flyingsilverfin) [reply database/RocksProperties.java:306]: Checked :)

**Pull request merged**